### PR TITLE
Adding caravan forcing

### PR DESCRIPTION
**Pull request by Daafip**
Hi All,

As this is my first contribution would like some feedback if possible. 

Was unsure on the naming convention, `generate` seems to be more linked to the esmvaltool, thus went for `retrieve`.

Runs for me when I run:

```py
from ewatercycle.forcing import sources
path = Path.cwd()
forcing_path = path / "Forcing"
experiment_start_date = "1997-08-01T00:00:00Z"
experiment_end_date = "2005-09-01T00:00:00Z"
HRU_id = 1022500

camels_forcing = sources['LumpedCaravanForcing'].retrieve(start_time = experiment_start_date,
                                                          end_time = experiment_end_date,
                                                          directory = forcing_path / "Camels",
                                                          basin_id = f"camels_0{HRU_id}",
                                                          variables = ('streamflow','potential_evaporation_sum'),
                                                          )
```
Discussions on the topic: #398 

Still todo: 
- [x] remove geopandas
- [x] remove wget
- [x] add documentation
- [x] add meaningfull example
- [x] Rename variables to correct ESMValTool 
- [x] Add test (using mocking)
- [x] fix failing test: `test_util.test_merge_esmvaltool_datasets`

### CODE REVIEW

**Comment by Daafip** [comment]
Not too sure how to complete renaming from ERA-5 to ESMValTool compatible variable names. 
Do you have any insight on how to do that @BSchilperoort?

Edit: for now just supplied a rename dictionary as in discussion

File: src/ewatercycle/_forcings/caravan.py
Comment (BSchilperoort) [inline src/ewatercycle/_forcings/caravan.py]: The code formatter doesn't like this.

You can run it with "pre-commit":
- `pip install pre-commit`
- `pre-commit run -a`

Usually it only looks at staged files, but with the `-a` flag you can run it on all files. This should also fix the CI.

File: src/ewatercycle/_forcings/caravan.py
Comment (BSchilperoort) [inline src/ewatercycle/_forcings/caravan.py]: It looks a lot like the other forcings now, which is nice. That should make it easier to swap this forcing with, e.g. Makkink, or the other way around.

File: src/ewatercycle/_forcings/caravan.py
Comment (BSchilperoort) [inline src/ewatercycle/_forcings/caravan.py]: I recently changed the code so the shapefile is copied along and stored next to the forcing files. 

That is part of the `.save` method. I think it works slightly differently here 🤔 . Or does it work the exact same?

File: src/ewatercycle/_forcings/caravan.py
Comment (BSchilperoort) [inline src/ewatercycle/_forcings/caravan.py]: I am in doubt about `retrieve` vs. `generate`...

`generate` would be better for consistency.
`retrieve` is more descriptive of what this method does.

At the least, you perhaps should make `.generate` raise a better error, at the moment you don't overwrite the default forcing's method even though that does contain an implementation.

Comment (BSchilperoort) [reply src/ewatercycle/_forcings/caravan.py]: @Peter9192 what is your opinion on this?

File: src/ewatercycle/_forcings/caravan.py
Comment (BSchilperoort) [inline src/ewatercycle/_forcings/caravan.py]: Did some of the basins miss lat and lon values? I expected all of them to have a proper value... Otherwise we should fix the input data.

It'd be better if it was correct there already instead of needing to fix it here.

File: src/ewatercycle/_forcings/caravan.py
Comment (BSchilperoort) [inline src/ewatercycle/_forcings/caravan.py]: I love python's sets 😄

Comment (BSchilperoort) [reply src/ewatercycle/_forcings/caravan.py]: I don't think you need it to be a tuple. You can iterate over a set as well.
Also, I believe `.intersection()` turns the argument into a set already.
```suggestion
        properties = set(variables).intersection(PROPERTY_VARS)
```

File: src/ewatercycle/_forcings/caravan.py
Comment (BSchilperoort) [inline src/ewatercycle/_forcings/caravan.py]: I think this should do the same, but is more readable:
```suggestion
        non_property_vars = set(variables) - properties
        variable_names= non_propery_vars.intersection(RENAME_ERA5.keys())
```

File: src/ewatercycle/_forcings/caravan.py
Comment (BSchilperoort) [inline src/ewatercycle/_forcings/caravan.py]:
```suggestion
        variables = ([RENAME_ERA5[var] for var in variable_names])
```

File: src/ewatercycle/_forcings/caravan.py
Comment (BSchilperoort) [inline src/ewatercycle/_forcings/caravan.py]: You can move this to the top. There won't be a distributed caravan forcing, so you can just inherit in the class containing the implementations.

Also, as there won't be a DistributedCaravanForcing, it might be better to just name it `CaravanForcing`.

File: src/ewatercycle/_forcings/caravan.py
Comment (BSchilperoort) [inline src/ewatercycle/_forcings/caravan.py]: You should wrap this in a context manager so `.close` is called on the response.

```py
with requests.get(...) as result:
    ...
```

File: src/ewatercycle/_forcings/caravan.py
Comment (BSchilperoort) [inline src/ewatercycle/_forcings/caravan.py]: Perhaps you can refactor this to a separate function to make the main function more readable.

**Review by BSchilperoort — CHANGES_REQUESTED** [review]
Hi David, nice work! I left some comments for possible improvements to the code. Feel free to ask a review again when it's ready!

**Comment by BSchilperoort** [comment]
Also, it would be nice if you could add tests for this forcing. (without downloading anything). If you're unsure on how to do that we can discuss it at some point

File: src/ewatercycle/_forcings/caravan.py
Comment (BSchilperoort) [reply src/ewatercycle/_forcings/caravan.py]: See https://github.com/WAM2layers/WAM2layers/pull/368#discussion_r1559049999 for some more in-depth info

File: src/ewatercycle/_forcings/caravan.py
Comment (Daafip) [reply src/ewatercycle/_forcings/caravan.py]: Yup, might also be nice to also have the seperate Camels data sets as these do vary slightly. (But thats more work for another time)

File: src/ewatercycle/_forcings/caravan.py
Comment (Daafip) [reply src/ewatercycle/_forcings/caravan.py]: Good point, will change that

File: src/ewatercycle/_forcings/caravan.py
Comment (Daafip) [reply src/ewatercycle/_forcings/caravan.py]: They all have lat/lon, was my first way of fixing the issue. Then i just moved the variables. Will remove the redundant code.

File: src/ewatercycle/_forcings/caravan.py
Comment (Daafip) [reply src/ewatercycle/_forcings/caravan.py]: Thanks, that cleans it up nicely

**Comment by RolfHut** [comment]
regarding conventions: from a didactic point of view in tutorial notebooks I suggest to always use:
```python
import ewatercycle.forcing

camels_forcing = ewatercycle.forcing.sources['LumpedCaravanForcing'].retrieve(start_time = experiment_start_date,
                                                          end_time = experiment_end_date,
                                                          directory = forcing_path / "Camels",
                                                          basin_id = f"camels_0{HRU_id}",
                                                          variables = ('streamflow','potential_evaporation_sum'),
                                                          )
```
Novice users quickly loose sight of which function belongs to which library when using ```from ewatercycle.forcing import sources```. This might load a bit more than you would ideally want, but communicates more clearly. In code 'under the hood' I'm more than fine with the shorter version.

File: src/ewatercycle/_forcings/caravan.py
Comment (Daafip) [reply src/ewatercycle/_forcings/caravan.py]: Made `.generate` now raises an `NotImplementedError` directing you to `retrieve`.
Retrieve makes more sense to me as it also makes it clear to the user it's an already prepared dataset. 
But interested to see what Peter says.

File: src/ewatercycle/_forcings/caravan.py
Comment (Daafip) [reply src/ewatercycle/_forcings/caravan.py]: Still getting issues with CI:
```
nbqa-isort...............................................................Failed
..
ModuleNotFoundError: No module named 'pkg_resources'
```
 and 
```
nbqa-flake8..............................................................Failed
...
AttributeError: 'EntryPoints' object has no attribute 'get'
```

**Comment by Daafip** [comment]
Made the suggested changes Rolf, Example Notebook can be found here: 
https://gist.github.com/Daafip/ac1b030eb5563a76f4d02175f2716fd7

**Comment by Daafip** [comment]
Still facing three issues, maybe you could help with that Bart?: 
- [x] test_utipl.test_merge_esmvaltool_datasets fails: . If i run the test locally in my IDE I don't have this problem but on the git tests I Do...? 
```
FAILED tests/src/test_util.py::test_merge_esmvaltool_datasets - AssertionError: assert 'height' in {'cell_methods': 'day_of_year: year: mean', 'long_name': 'Near-Surface Air Temperature', 'standard_name': 'air_temperature', 'units': 'K'}
```
- [x] `MyPy` throws an error I dont understand in the SonarCloud analysis:
```
mypy.....................................................................Failed
-hook id: mypy
-exit code: 1

src/ewatercycle/_forcings/caravan.py:10: error: Library stubs not installed for "requests"  [import]
src/ewatercycle/_forcings/caravan.py:10: note: Hint: "python3 -m pip install types-requests"
src/ewatercycle/_forcings/caravan.py:10: note: (or run "mypy --install-types" to install all missing stub packages)
src/ewatercycle/_forcings/caravan.py:10: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
```
- [x] Not sure how to make tests not download

**Comment by BSchilperoort** [comment]
> test_utipl.test_merge_esmvaltool_datasets fails: . If i run the test locally in my IDE I don't have this problem but on the git tests I Do...?

That's odd. The "height" attribute should be in there. Perhaps it's an environment/versioning issue, and the environment is newly created on Github Actions for the test. So you don't see the error as you don't have all the latest versions of packages.

> MyPy throws an error I dont understand in the SonarCloud analysis:

Probably also due to a version change. You can add the `types-request` package to the setup.cfg file below `types-PyYAML`. That should fix it.

> Not sure how to make tests not download

You would have to mock/patch some of the code. For examples see https://github.com/eWaterCycle/ewatercycle/blob/b4ea453727aef571dbc18459d1804c8f86c75ff0/tests/src/test_container.py#L177

You'd mock this line in your code `ds = xr.open_dataset(f"{OPENDAP_URL}{dataset}.nc")`, with `mock.patch("xarray.Dataset') as ...`.
You can then set the return value to return a dataset you load from a small testing file. And assert that the mocked function was called with the correct arguments.

**Comment by Daafip** [comment]
> That's odd. The "height" attribute should be in there. Perhaps it's an environment/versioning issue, and the environment is newly created on Github Actions for the test. So you don't see the error as you don't have all the latest versions of packages.

Still dont really understand why this fails, I think it might be a versioning issue, played around a bit in a notebook (https://gist.github.com/Daafip/3a4f8650f499e9a2b68fc187ef7d73d7) and it only worked if I loaded the function seperately. This is in a new environment using the eWaterCycle/ewatercycle source code and the latest pip release.  Unsure how to fix this though..

> SonarCloud analysis

Fixed mypy, now have an issue with the SonarCloud token, but i can't acces the repo settings..

"Anyone with collaborator access to this repository can use these secrets and variables for actions. They are not passed to workflows that are triggered by a pull request from a fork." - any way to fix this? 

> You would have to mock/patch some of the code. For examples see

 Ah okay, Will look at mokking tests.

File: src/ewatercycle/_forcings/caravan.py
Comment (Peter9192) [reply src/ewatercycle/_forcings/caravan.py]: From reading the conversation only I think I'm okay with the proposed solution :-)

**Comment by Daafip** [comment]
Added a basic test now, but took me quite a bit of time. 
Would you have time to take a look at the issues im having with SonarCloud & the other test @BSchilperoort?

**Comment by RolfHut** [comment]
apart from the technical comments, a more hydrological point: we do have to make sure in the documentation that people are aware that the Caravan dataset is not "just" the Camels grouped together, but rather it is new (ERA) data derived for the shape files in the Camels datasets. See https://egusphere.copernicus.org/preprints/2024/egusphere-2024-864/ for a ongoing discussion in the hydrological community about this.

**Comment by BSchilperoort** [comment]
> Added a basic test now, but took me quite a bit of time. Would you have time to take a look at the issues im having with SonarCloud & the other test @BSchilperoort?

Fixed the isort and black issues. I now see that there's the "end-of-file-fixer" which (incorrectly!) is trying to fix those shapefile file. I did not find any way to configure it, so I removed it from the pre-commit setup. Black already adds a newline to the end of .py files anyway.

File: tests/src/base/test_forcing.py
Comment (BSchilperoort) [inline tests/src/base/test_forcing.py:282]: I don't think this mock is not being used below. Have a look at the test_container.py file.

The fixture needs to be an input argument for the test function.

By adding `mocked_thing.assert_called_once_with(...` you can make sure that the mock was correctly used.

**Comment by BSchilperoort** [comment]
Sonarcloud is still failing due to "Set the SONAR_TOKEN env variable.". However, in the repo's secrets this is correctly set and it used to work fine.

No idea how to fix this. There doesn't seem to be a new version out so it's completely unclear to me why the workflow stopped working. Three weeks ago the it still ran successfully. 

Any ideas @sverhoeven ?

**Comment by Daafip** [comment]
> Sonarcloud is still failing due to "Set the SONAR_TOKEN env variable.". However, in the repo's secrets this is correctly set and it used to work fine.
> 
> No idea how to fix this. There doesn't seem to be a new version out so it's completely unclear to me why the workflow stopped working. Three weeks ago the it still ran successfully.
> 
> Any ideas @sverhoeven ?

I think it might be because I forked, can you create a new branch in the repo (i dont have the rights to do so) and see if that fixes it?

File: tests/src/base/test_forcing.py
Comment (Daafip) [reply tests/src/base/test_forcing.py:282]: Ah, will look into this

**Comment by BSchilperoort** [comment]
> > Sonarcloud is still failing due to "Set the SONAR_TOKEN env variable.". However, in the repo's secrets this is correctly set and it used to work fine.
> > No idea how to fix this. There doesn't seem to be a new version out so it's completely unclear to me why the workflow stopped working. Three weeks ago the it still ran successfully.
> > Any ideas @sverhoeven ?
> 
> I think it might be because I forked, can you create a new branch in the repo (i dont have the rights to do so) and see if that fixes it

I think you're completely correct with that! Good thinking :+1:

I won't bother making a new branch on this repo, pretty sure that's the cause of the problem. I have given you write access though, so for a next PR you can make a branch on the repo.

File: tests/src/base/forcing_files/test_caravan_file.nc
Comment (BSchilperoort) [inline tests/src/base/forcing_files/test_caravan_file.nc:1]: If these are real Caravan netCDF and shape files it might be best to add a readme with a license statement.

File: tests/src/base/test_forcing.py
Comment (Daafip) [reply tests/src/base/test_forcing.py:282]: It indeed wasn't actually mocking it properly using  `with mock.patch("xarray.open_dataset")`
If i use `with mock.patch("xarray.open_dataset")...`: it calls it correctly and opens the dataset, but also mocks `open_dataset` in the `caravan_forcing.to_xarray()` which then break things...
Will keep trying to fix it

Comment (Daafip) [reply tests/src/base/test_forcing.py:282]: ah fixed it, https://changhsinlee.com/pytest-mock/ this is a useful resource

**Comment by Daafip** [comment]
> That's odd. The "height" attribute should be in there. Perhaps it's an environment/versioning issue, and the environment is newly created on Github Actions for the test. So you don't see the error as you don't have all the latest versions of packages.

Still don't understand why this test fails...

Otherwise I think its mostly there

**Comment by sverhoeven** [comment]
Can you run pre-commit locally? `pip install pre-commit` + `pre-commit run --all-files`, you can ignore the flake8 warnings, but errors should be fixed.

File: .pre-commit-config.yaml
Comment (sverhoeven) [inline .pre-commit-config.yaml:10]: Instead of removing hook can shape files be excluded 
see https://pre-commit.com/index.html#regular-expressions

Comment (Daafip) [reply .pre-commit-config.yaml:10]: thanks, fixed that now. Still failing on the test but once we mere that PR, should be fixed

**Comment by BSchilperoort** [comment]
Hi David, we found the test issue and fixed it in #410 . Once that's merged to main that should resolve the problem you encounter here.

On second thought I do think that it would be better to use the `generate` method to generate the forcing. You can add extra kwargs (i.e. basin_id, and just leave the `shape` one unused).

From the converage output in the CI I see this;
```
src/ewatercycle/_forcings/caravan.py                   85     28     32      3    60%
```

Generally we want 80% coverage on the code (sonarcloud analysis). I think if you add tests for the `extract_basin_shapefile` function that you might be there already.

After these things I think this PR is ready to merge! :+1:

**Comment by Daafip** [comment]
> I think if you add tests for the extract_basin_shapefile

Done! That Bumps it up to 83%
`src/ewatercycle/_forcings/caravan.py                   84     11     32      3    83%`

> On second thought I do think that it would be better to use the generate method to generate the forcing. You can add extra kwargs (i.e. basin_id, and just leave the shape one unused).

Refctored this now, so all ready! 

Pytest fails on the other test, sonar cloud is stuck on the token but should both be fine.

Thanks for all the input & feedback @BSchilperoort @sverhoeven!

**Comment by Daafip** [comment]
> apart from the technical comments, a more hydrological point: we do have to make sure in the documentation that people are aware that the Caravan dataset is not "just" the Camels grouped together, but rather it is new (ERA) data derived for the shape files in the Camels datasets. See https://egusphere.copernicus.org/preprints/2024/egusphere-2024-864/ for a ongoing discussion in the hydrological community about this.

To get back to you on this  @RolfHut, In the data set attributes under history it does mention the data source. This is now correctly copied over to the new datas set. If the user looks at the (meta)data, they should see its ERA5

File: src/ewatercycle/_forcings/caravan.py
Comment (BSchilperoort) [inline src/ewatercycle/_forcings/caravan.py:144]: Can you add a check that gives the user a clear input error if the basin ID is missing?

Comment (BSchilperoort) [reply src/ewatercycle/_forcings/caravan.py:144]: Once this is fixed and you've merged from main (failing test is fixed in there), I'll approve and merge the PR.

Can you also update the changelog?

Comment (Daafip) [reply src/ewatercycle/_forcings/caravan.py:144]: Ah forgot to remove line 139, the `basin_id` is now passed as a kwarg rather than using the dataset variable. 

Will merge from main and update changelog

File: CHANGELOG.md
Comment (BSchilperoort) [inline CHANGELOG.md]:
```suggestion
- Added LumpedCaravan forcing, which allows for the retrieval of single basins from the Caravan dataset, rather than downloading the whole dataset ([#407](https://github.com/eWaterCycle/ewatercycle/pull/407)).
```

File: src/ewatercycle/_forcings/caravan.py
Comment (BSchilperoort) [inline src/ewatercycle/_forcings/caravan.py:144]:
```suggestion
        if "basin_id" not in kwargs:
            msg = "You have to specify a basin ID to be able to generate forcing from Caravan."
            raise InputError(msg)
        basin_id = kwargs["basin_id"]
```

File: src/ewatercycle/_forcings/caravan.py
Comment (BSchilperoort) [inline src/ewatercycle/_forcings/caravan.py]: Can you add where the user can lookup the basin ID?

**Review by BSchilperoort — APPROVED** [review]
Nice work! Sorry that it was blocked a bit by the failing test+sonarcloud problems, but now it's ready to merge (after that 1 last comment above, and pre-commit haha) :smile: 

Edit: InputError isn't a buildin exception. You can pick a different appropriate one. This is why mypy fails.

File: src/ewatercycle/_forcings/caravan.py
Comment (Daafip) [reply src/ewatercycle/_forcings/caravan.py]: See this is where geopandas would be nice :p, but will add a function which lets the user explore the dataset(ids).

Comment (BSchilperoort) [reply src/ewatercycle/_forcings/caravan.py]: I was hoping more for something like a webpage listing the basin_id's together with river/basin names, country, etc. I guess that doesn't exist?

Comment (Daafip) [reply src/ewatercycle/_forcings/caravan.py]: Nope, but what i do now is list the datasets in a seperate function. That way the user can get the dataset as netcdf and a list of basin ID's

Comment (Daafip) [reply src/ewatercycle/_forcings/caravan.py]: For advanced users this will do, for more novice users choosing a single catchment: likely they will have to download the combinedshapefile, load this into GIS and pick one

Comment (Daafip) [reply src/ewatercycle/_forcings/caravan.py]: > I was hoping more for something like a webpage listing the basin_id's together with river/basin names, country, etc. I guess that doesn't exist?

would be pretty easy to make an interactive folium/leaflet map out of this actually! Similar to [this](https://nbviewer.org/github/Daafip/KNMI-Folium-Map/blob/main/Hourly%20%26%20daily%20station%20data%20KNMI%202023-03-21.html)

Comment (Daafip) [reply src/ewatercycle/_forcings/caravan.py]: Lets continue this discussion in issues: #398

**Pull request merged**
